rfcs/2021-05-23-groups2.md

```diff
@@ -0,0 +1,134 @@
+  - integrity check message mush contain a signature per member, so it would be substantially larger.
+  - agents must use the same verification key for all members in the group, complicating the group and connections management.
+  - in the same way as in the first scenario, some delivery notifications can be arbitrarily delayed.
+
+With either approach, each agent should probably track all the receipts of all messages from all agents. 
+
+Overall the first approach seems better. It shows who received which messages and the only case where the lack of integrity would be reported if the same message is different or some messages are skipped to some recipients (i.e. they broadcast integrity violation for some of the senders).
+
+## Agent commands and messages syntax
+
+- command `G:gId? NEW` - create group (response is `G:gId OK`, or `ERR` if this group already exists)
+- command `C:cId INTRO G:gId gInfo` - invite existing connection to a group
+- message `C:cId REQ G:invID gInfo` - invitation to join the group
+- command `G:gId? ACPT G:invId` - accept invitation (response is `G:gId OK`)
+- message `G:gId CON C:cId` - 2 connections created with some group member (both for group and direct messages)
+- message `G:gId MEM [C:cId]` - connection created with all group members for a given member or current client
+- command `G:gId SEND msg` - send message to group
+- message `G:gId SENT [C:cId] msgId` - notification that the message is sent and its internal ID
+- message `G:gId MSG C:cId msgdata` - received group message from cId, msgdata is the same set of parameters as in `MSG`
+- command `G:gId ACK msgId` - acknowledge message reception by the client
+- message `G:gId RCVD [C:cId] msgId status` - message delivery notification
+- command `G:gId LEAVE` - leave the group
+- message `G:gId LEFT [C:cId]` - connection cId left the group
+- command `G:gId REM C:cId` - remove group member (response is `G:gId OK`, followed by `REMD` notification)
+- message `G:gId REMD C:cId [C:cId]` - member removed (who, by whom - if it's not the current user)
+- message `G:gId OUT C:cId` - you are removed
+- message `G:gId EMPTY` - all members left the group and it is now empty
+- command `G:gId DEL` - delete the group (response is `G:gId OK`)
+- message `G:gId DELD [C:cId]` - group deleted (by whom, if it's not the current user)
+- command `G:bId LS` - list connections in broadcast, response is `MEM space_separated_connections`
+- message `G:bId MS space_separated_connections`
+
+## Agent message envelopes syntax (group-specific)
+
+- `GROUP C:mid G:inv gInfo` - invitation to join the group
+- `MEM C:mid` - confirmation that member connected to all members
+- `LEFT` - notification that member left the group
+- `OUT` - you are removed from the group
+- `REM C:mid` - remove member mid from the group
+- `REMD C:mid` - confirmation that member is removed
+- `DEL` - group is deleted
+- `DELD` - confirmation that group is deleted
+
+## Protocol costs
+
+- Adding a member:
```

Comment:
What are these magic numbers '1', '4', '6'

Reply:
the number of messages transmitted to connect members, 4 is from introduction protocol, 6 is total.